stan/math/rev/core/var.hpp

```diff
@@ -25,36 +23,49 @@ static void grad(vari* vi);
  * an arena-based memory manager scoped to a single gradient
  * calculation.
  *
- * A var is constructed with a double and used like any
+ * A var is constructed with a type `T` and used like any
  * other scalar.  Arithmetical functions like negation, addition,
  * and subtraction, as well as a range of mathematical functions
  * like exponentiation and powers are overridden to operate on
  * var values objects.
+ * @tparam T An Arithmetic type.
  */
-class var {
+template <typename T, typename = void>
+class var_value {};
+
+template <typename T>
+class var_value<T, require_floating_point_t<T>> {
  public:
-  // FIXME: doc what this is for
-  using Scalar = double;
+  // The internal value_type is always a floating point type
+  using value_type = std::decay_t<T>;
+  using vari_type = vari_value<value_type>;
+  using vari_pointer = vari_type*;
+  template <typename K>
+  using require_vari_convertible_t
+      = require_t<std::is_convertible<vari_value<K>*, vari_pointer>>;
+  template <typename K>
+  using require_not_vari_convertible_t = require_t<
+      bool_constant<!std::is_convertible<vari_value<K>*, vari_pointer>::value>>;
 
   /**
    * Pointer to the implementation of this variable.
    *
    * This value should not be modified, but may be accessed in
-   * <code>var</code> operators to construct <code>vari</code>
+   * <code>var</code> operators to construct `vari_value<T>`
    * instances.
    */
-  vari* vi_;
+  vari_type* vi_;
 
   /**
-   * Return <code>true</code> if this variable has been
+   * Return `true` if this variable has been
    * declared, but not been defined.  Any attempt to use an
    * undefined variable's value or adjoint will result in a
    * segmentation fault.
    *
    * @return <code>true</code> if this variable does not yet have
    * a defined variable.
    */
-  bool is_uninitialized() { return (vi_ == static_cast<vari*>(nullptr)); }
+  bool is_uninitialized() { return (vi_ == static_cast<vari_type*>(nullptr)); }
```

Comment:
I think we don't need to cast here.
```suggestion
  bool is_uninitialized() { return (vi_ == nullptr); }
```